akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala

```diff
@@ -237,7 +236,8 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with LogHelper
     def handle(event: StreamFrameEvent): StreamState =
       if (event.isInstanceOf[ParsedHeadersFrame] && streamStates.size > settings.maxConcurrentStreams) {
         // When trying to open a new Stream, if that op would exceed the maxConcurrentStreams, then refuse the op
-        resetStream(event.streamId, ErrorCode.REFUSED_STREAM)
+        debug("Peer trying to open stream that would exceed `maxConcurrentStreams`, refusing stream")
+        multiplexer.pushControlFrame(RstStreamFrame(event.streamId, ErrorCode.REFUSED_STREAM))
```

Comment:
hmm, so `resetStream` is only to be called from outside the state machine, while inside the state machine you have to push the RST 'yourself'?

Reply:
Unfortunately, yes. The alternative would be either to introduce an effect system to make side-effects more explicit, or to introduce post-actions that automatically reset a stream if it moved to `Closed` state. I might experiment with one of them but for now this seems to be the right way.